### PR TITLE
Fix managed storage synchronization issue

### PR DESCRIPTION
**Pull request by gronerl**
* Fix performance issue where setitem in GPU storage always triggers host-to-device copy.
* Improve tracking of storages involved in a stencil, that might warrant a device synchronization.

@eddie-c-davis You originally introduced the change in `__setitem__` where `value.data` was used in `cp.asarray`. That actually always causes a host-to-device copy, can you please verify that this PR works for you?

### CODE REVIEW

**Comment by gronerl** [comment]
bors try

File: src/gt4py/storage/storage.py [hunk 1]
```diff
@@ -300,41 +303,48 @@ def __getitem__(self, item):
     def __setitem__(self, key, value):
         if hasattr(value, "__cuda_array_interface__"):
             gpu_view = storage_utils.gpu_view(self)
-            gpu_view[key] = cp.asarray(value.data)
-            self.device_to_host(True)
+            self.device_to_host()
+            if isinstance(value, GPUStorage):
+                value.device_to_host()
```
Comment (fthaler) [inline src/gt4py/storage/storage.py:308]: I don’t understand why the two `device_to_host` calls are necessary. Why should setitem from GPU to GPU storage ever trigger a sync? The sync is only necessary on the next access on the host, right? So for me a call to `self._set_device_modified()` seems to be the right thing here, but maybe I miss something?

File: src/gt4py/storage/storage.py [hunk 2]
```diff
@@ -259,13 +261,14 @@ def _construct(cls, backend, dtype, default_origin, shape, alignment, layout_map
         )
         obj = field.view(_ViewableNdarray)
         obj = obj.view(GPUStorage)
+        print("construct", raw_buffer[-1:].data.ptr)
```
Comment (fthaler) [inline src/gt4py/storage/storage.py:264]: Please remove.

**Review by fthaler — CHANGES_REQUESTED** [review]

**Comment by eddie-c-davis** [comment]
Hello @gronerl, thank you, I shall test this with `fv3core`.

File: src/gt4py/storage/storage.py (hunk 1 above)
Comment (gronerl) [reply src/gt4py/storage/storage.py:308]: Basically, if there's a stencil(s) that may still have running kernels,  the setitem may overwrite the same buffer and you get a race condition. The `device_to_host` is a misnomer in this context, all it does is to see if any of the involved storages depends on a recent kernel and waits for it to finish if necessary. Why not only one device_to_host: Previously it would just always force a synchronization, now we only do this if any of the two involved storages depends on a kernel.

Comment (fthaler) [reply src/gt4py/storage/storage.py:307]: There is no race condition as long as the kernel and the copy run on the same stream, which I assume is always the case (we always use the default stream afaik).

Comment (gronerl) [reply src/gt4py/storage/storage.py:307]: Note that the cupy sync in device_to_host is not a storage sync that equalizes the buffers, it is actually a `cudaDeviceSynchronize`

Comment (fthaler) [reply src/gt4py/storage/storage.py:307]: I know and this is slow as it blocks everything… And as said, this is not required if everything runs on the same (default) stream.

Comment (gronerl) [reply src/gt4py/storage/storage.py:307]: Sorry, our messages crossed. I see what you mean now, true, the sync would actually have to go in the "else" clause. I'll fix it.

File: tests/test_unittest/test_storage.py [hunk 3]
```diff
@@ -940,11 +940,33 @@ def swap_stencil(
         shape=shape,
         managed_memory=True,
     )
-
+    q0_view = q0[3:, 3:, 3:]
+    print(id(q0))
+    print(id(q1))
+    print(id(q0_view))
```
Comment (jdahm) [inline tests/test_unittest/test_storage.py:946]: Remove before merging?

Comment (gronerl) [reply tests/test_unittest/test_storage.py]: thanks, removed.

**Comment by eddie-c-davis** [comment]
Hello @gronerl, I tested this on `fv3core` and received an error in some code that copies corner values between two storages. This happens outside of a stencil. Here is a small reproducer,
```python
    backend = "gtcuda"
    shape = (6, 6, 6)

    u = gt_store.from_array(
        cp.zeros(shape),
        backend=backend,
        dtype=np.float64,
        default_origin=(0, 0, 0),
        shape=shape,
        managed_memory=True,
    )

    v = gt_store.from_array(
        cp.ones(shape),
        backend=backend,
        dtype=np.float64,
        default_origin=(0, 0, 0),
        shape=shape,
        managed_memory=True,
    )

    u[0, 2, :] = -v[2, 5, :]
```
The error is:
```python
E   cupy_backends.cuda.api.runtime.CUDARuntimeError: cudaErrorInvalidDevice: invalid device ordinal

cupy_backends/cuda/api/runtime.pyx:261: CUDARuntimeError
```
The error vanishes when `cp.asarray(value)` is changed back to `cp.asarray(value.data)`.  I have not had time to dig any further into the error.

**Comment by gronerl** [comment]
Thanks for testing the PR. 

So the problem is caused by the `-` sign. This triggers the `np.negative` ufunc which are generally not supported. What happens is that you get a non-GPU array underlying the GPU storage, which you are then trying to assign to a GPU storage. If you put the `.data` attribute, a copy is forced to GPU and it works again. 

I'm Not sure how to deal with this. Ufuncs are not supported for a reason: It is in many cases impossible to make sure the resulting storages are compatible / meaningful for use in stencils. Is there a way to not ever use them in fv3core? Note that in-place augmented operations of the likes of `-=` should work properly.

**Comment by eddie-c-davis** [comment]
Excellent catch, thank you! Yes, I shall refactor the `fv3core` code and test again.

**Review by eddie-c-davis — APPROVED** [review]
Looks great to me, our tests are now passing.

**Comment by gronerl** [comment]
I removed the unneeded syncs. Since @fthaler is away, can e.g. @havogt check if this makes sense?

**Review by havogt — APPROVED** [review]
Looks good, but didn't do a deep review.
